torchvision/utils.py

```diff
@@ -229,49 +229,70 @@ def draw_segmentation_masks(
-        raise TypeError(f"Tensor expected, got {type(image)}")
-    elif image.dtype != torch.uint8:
-        raise ValueError(f"Tensor uint8 expected, got {image.dtype}")
+        raise TypeError(f"The image must be a tensor, got {type(image)}")
+    elif image.dtype not in (torch.uint8, torch.float):
+        raise ValueError(f"The image dtype must be uint8 or float, got {image.dtype}")
     elif image.dim() != 3:
         raise ValueError("Pass individual images, not batches")
     elif image.size()[0] != 3:
         raise ValueError("Pass an RGB image. Other Image formats are not supported")
+    if masks.ndim == 2:
+        masks = masks[None, :, :]
+    if masks.ndim != 3:
+        raise ValueError("masks must be of shape (H, W) or (batch_size, H, W)")
+    if masks.dtype != torch.bool:
+        raise ValueError(f"The masks must be of dtype bool. Got {masks.dtype}")
+    if masks.shape[-2:] != image.shape[-2:]:
+        raise ValueError(f"The image and the masks must have the same height and width")
 
     num_masks = masks.size()[0]
-    masks = masks.argmax(0)
+    if colors is not None and num_masks > len(colors):
+        raise ValueError(f"There are more masks ({num_masks}) than colors ({len(colors)})")
 
     if colors is None:
-        palette = torch.tensor([2 ** 25 - 1, 2 ** 15 - 1, 2 ** 21 - 1])
-        colors_t = torch.as_tensor([i for i in range(num_masks)])[:, None] * palette
-        color_arr = (colors_t % 255).numpy().astype("uint8")
-    else:
-        color_list = []
-        for color in colors:
-            if isinstance(color, str):
-                # This will automatically raise Error if rgb cannot be parsed.
-                fill_color = ImageColor.getrgb(color)
-                color_list.append(fill_color)
-            elif isinstance(color, tuple):
-                color_list.append(color)
-
-        color_arr = np.array(color_list).astype("uint8")
-
-    _, h, w = image.size()
-    img_to_draw = Image.fromarray(masks.byte().cpu().numpy()).resize((w, h))
-    img_to_draw.putpalette(color_arr)
-
-    img_to_draw = torch.from_numpy(np.array(img_to_draw.convert('RGB')))
-    img_to_draw = img_to_draw.permute((2, 0, 1))
-
-    return (image.float() * alpha + img_to_draw.float() * (1.0 - alpha)).to(dtype=torch.uint8)
+        colors = _generate_color_palette(num_masks)
+
+    if not isinstance(colors, list):
+        colors = [colors]
+    if not isinstance(colors[0], (tuple, str)):
+        raise ValueError("colors must be a tuple or a string, or a list thereof")
+    if isinstance(colors[0], tuple) and len(colors[0]) != 3:
+        raise ValueError("It seems that you passed a tuple of colors instead of a list of colors")
+
+    out_dtype = image.dtype
+
+    colors_ = []
+    for color in colors:
+        if isinstance(color, str):
+            color = ImageColor.getrgb(color)
+        color = torch.tensor(color, dtype=out_dtype)
+        if out_dtype == torch.float:
+            color /= 255
```

Comment:
I wonder if we have to support anything other than uint8 here, given that this method is used for visualization. To be specific, the only issue I have with this is that you are being forced to assume that the colour needs to be between 0-1, hence divide by 255. Thoughts?

Reply:
I too thought (I will have a thorough look tommorrow!). The other utility `draw_bounding_boxes` return `uint8` dtype only.

Reply:
The current interface takes a uint8 image as input and outputs a float image. This means that you can't call `draw_segmentation_masks` on that output again, which isn't super practical.

Also, the models accept float images, not uint8 images. If you look at the current example https://pytorch.org/vision/master/auto_examples/plot_visualization_utils.html#sphx-glr-auto-examples-plot-visualization-utils-py we're forced to have 2 copies of each image: one float and one for uint8. This isn't optimal either.

Supporting both floats and uint8 and have the dtypes "pass through" solves both of these issues for little maintenance cost - I wrote tests that ensure both dtypes yield the same results.

Regarding consistency, I'm planning on modifying draw_bounding_boxes to have the same behavior.

Reply:
> Also, the models accept float images, not uint8 images.

It is true the models receive float images that are also typically Normalized. These are not the versions of the images that you will use for visualization though. Most models scale to [0-1] while others recently require scaling to [-1,1] (see the SSD models). So at the end the user will need to keep around the non-normalized `uint8` version to visualize it.

An alternative course of action might be to continue handling all types of images but avoid any assumption over the color scales. You could do that by creating a palette only if the image is uint8 and expect the user to provide a valid list of colors on their right scales otherwise. Happy to discuss this more to unblock your work.

Reply:
> It is true the models receive float images that are also typically Normalized. These are not the versions of the images that you will use for visualization though

According to https://github.com/pytorch/vision/pull/3774#discussion_r629332764, Mask-RCNN and Faster-RCNN don't require normalization, so one could use the same non-normalized float images in this case I think.

IIUC, the main concern seems to be the assumption that if an image is a float, then we expect it to be in [0, 1]. [According to our docs](https://pytorch.org/vision/master/transforms.html), this assumption holds for the transforms already: 

> The expected range of the values of a tensor image is implicitely defined by the tensor dtype. Tensor images with a float dtype are expected to have values in [0, 1). Tensor images with an integer dtype are expected to have values in [0, MAX_DTYPE] where MAX_DTYPE is the largest value that can be represented in that dtype.

It also holds when calling `F.convert_image_dtype(img, torch.float)` and all our conversion utilities as far as I can tell, so it seems to be a reasonable assumption.

I'll add a comment in the docstring to clarify that. The worst thing that can happen if the user passes a float image that's not in [0, 1] is that the colors will be slightly off

Reply:
The normalization for Segmentation happens outside of the models:

https://github.com/pytorch/vision/blob/e35793a1a4000db1f9f99673437c514e24e65451/references/segmentation/presets.py#L25-L29

Because the same applies on Classification and because we are in the process of redesigning the transforms for Detection, it's not a given that Normalization will continue to happen within those models. As of lately, the Transforms of Detection started looking a lot like the ones on classification that include the Normalization outside:
https://github.com/pytorch/vision/blob/e35793a1a4000db1f9f99673437c514e24e65451/references/classification/presets.py#L29-L34
https://github.com/pytorch/vision/blob/e35793a1a4000db1f9f99673437c514e24e65451/references/detection/presets.py#L12-L18


I understand that your motivation is to avoid keeping around both the uint8 and the float versions of the picture. Nevertheless throwing away the integer image means you will have to undo the [0,1] scaling before passing the picture to PIL for vizualization. I think this is more inconvinient for the user than keeping a reference to the `uint8` image.


Finally it is true that ToTensor() silently rescales the pixels values to [0,1] by dividing by 255. This assumption is used as you mentioned by some other transforms but it is kept contained. Exposing this detail to the visualization scripts means that it becomes even harder on the future to change it.


For the above reasons, I recommend not rely on assumptions that we might not apply soon. Your specific method does not have to handle colour normalization for the users. You can just remove that part and avoid introducing extra assumptions about the scales within your function.

Reply:
> Finally it is true that ToTensor() silently rescales the pixels values to [0,1] by dividing by 255. This assumption is used as you mentioned by some other transforms but it is kept contained. Exposing this detail to the visualization scripts means that it becomes even harder on the future to change it.

Are there plans to change the behaviour of `ToTensor()` and all of the transforms w.r.t. this assumption with float inputs?

Again, I don't think there's much risk in assuming that float images are in [0, 1] for this utility. Forcing users to explicitly pass colors when using floats is unpractical. And let's keep in mind that i) this `/= 255` division **only** affects the masks colors, not the input image color and ii) plotting the result masks an image that has been normalized (i.e. that isn't in 0-1) leads to terrible visualizations anyway, so I don't think anyone would do that:

For example with 2 images normalized with `F.normalize(batch, mean=(0.485, 0.456, 0.406), std=(0.229, 0.224, 0.225))`

![image](https://user-images.githubusercontent.com/1190450/118104564-35fa4e80-b3d3-11eb-9401-4bb11a6468b7.png)

Before they can visualize such image properly users would need to de-normalize it to map it into [0, 1], so they might as well do that before they use `draw_segmentaiton_masks`

Reply:
In other words: as far as I can tell, removing the assumption that float images are in [0, 1] for `draw_segmentation_masks` will not facilitate any use-case, and it won't make anything worse either.

OTOH, removing this assumption forces users to keep the uint8 image around, or it forces them to manually pass colors. By seamlessly allowing float images in [0, 1] users only need to keep the float image in [0, 1], and possibly the normalized image to pass to a model. They can discard the uint8 image.

Reply:
> Are there plans to change the behaviour of ToTensor() and all of the transforms w.r.t. this assumption with float inputs?

My point here is about keeping this assumption contained and not exposing it to other places of the library. Whether we should change the behaviour of ToTensor to remove the silent rescaling is something we could discuss for TorchVision v1.0.

>  plotting the result masks an image that has been normalized (i.e. that isn't in 0-1) leads to terrible visualizations anyway

This is precisely why the original versions of these methods supported only uint8 images and that's why we can't just throw away the uint8 versions of the images. The Segmentation models rescale and normalize the images outside of the models meaning you would have to undo them before vizualizing.

At this point I think it would be best not to merge the PR without containing the Gallery examples that would give us a clear view of the usage and the how corner-cases are handled. Perhaps @fmassa can weight-in here.

Reply:
> The Segmentation models rescale and normalize the images outside of the models meaning you would have to undo them before vizualizing.

Users **have to** convert their images to float in [0, 1] at some point.

> All pre-trained models expect input images normalized in the same way,
i.e. mini-batches of 3-channel RGB images of shape (3 x H x W),
where H and W are expected to be at least 224.
The images have to be loaded in to a range of [0, 1] and then normalized
using ``mean = [0.485, 0.456, 0.406]`` and ``std = [0.229, 0.224, 0.225]``.

So it's not just something contained in the transforms. It's pervasive throughout the library, and throughout the entire ecosystem. By supporting those images directly, we allow users to drop the uint8 images and only keep the float images (normalized and non-normalized versions).

Reply:
Still WIP but see also https://github.com/pytorch/vision/pull/3824/files#r631748970 to illustrate why and how we can drop references to uint8 images once we support floats in 0-1